// @vue-macros/setup-block@0.0.1/dist/chunk-XMPHL6QM.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/index.ts
var _unplugin = require('unplugin');
var _pluginutils = require('@rollup/pluginutils');
var _common = require('@vue-macros/common');

// src/core/index.ts

var _compilerdom = require('@vue/compiler-dom');
var transformSetupBlock = (code, id, lang) => {
  const s = new (0, _common.MagicString)(code);
  const node = _compilerdom.parse.call(void 0, code, {
    isNativeTag: () => true,
    isPreTag: () => true,
    getTextMode: ({ tag, props }, parent) => {
      if (!parent && tag !== "template" || tag === "template" && props.some(
        (p) => p.type === 6 && p.name === "lang" && p.value && p.value.content && p.value.content !== "html"
      )) {
        return 2;
      } else {
        return 0;
      }
    }
  });
  for (const child of node.children) {
    if (child.type === 1 && child.tag === "setup") {
      const hasLang = child.props.some((p) => p.name === "lang");
      let codegen = "script setup";
      if (!hasLang && lang) {
        codegen += ` lang="${lang}"`;
      }
      s.overwrite(
        child.loc.start.offset + 1,
        child.loc.start.offset + 6,
        codegen
      );
      s.overwrite(
        child.loc.end.offset - 6,
        child.loc.end.offset - 1,
        "script"
      );
    }
  }
  return _common.getTransformResult.call(void 0, s, id);
};

// src/index.ts
function resolveOption(options) {
  return {
    include: [_common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC],
    defaultLang: "ts",
    ...options
  };
}
var name = "unplugin-vue-setup-block";
var src_default = _unplugin.createUnplugin.call(void 0, (userOptions = {}) => {
  const options = resolveOption(userOptions);
  const filter = _pluginutils.createFilter.call(void 0, options.include, options.exclude);
  return {
    name,
    enforce: "pre",
    transformInclude(id) {
      return filter(id);
    },
    transform(code, id) {
      try {
        return transformSetupBlock(code, id, options.defaultLang);
      } catch (err) {
        this.error(`${name} ${err}`);
      }
    }
  };
});



exports.src_default = src_default;
