// @pureadmin/table@1.8.3/dist/index.es.js
var De = Object.defineProperty;
var xe = (e, i, o) => i in e ? De(e, i, { enumerable: !0, configurable: !0, writable: !0, value: o }) : e[i] = o;
var de = (e, i, o) => (xe(e, typeof i != "symbol" ? i + "" : i, o), o);
import * as X from "vue";
import { defineComponent as we, createVNode as L, Fragment as Oe, getCurrentInstance as je, onMounted as Te, nextTick as Ce, toRefs as Ie, computed as re, unref as v, withDirectives as _e, mergeProps as q, resolveDirective as ke, isVNode as Ne } from "vue";
import { ElTable as He, ElPagination as Ye, ElTableColumn as Fe } from "element-plus";
var Pe = { data: { type: Array, default: () => [] }, size: String, width: [String, Number], height: [String, Number], maxHeight: [String, Number], fit: { type: Boolean, default: !0 }, stripe: Boolean, border: Boolean, rowKey: [String, Function], showHeader: { type: Boolean, default: !0 }, showSummary: Boolean, sumText: String, summaryMethod: Function, rowClassName: [String, Function], rowStyle: [Object, Function], cellClassName: [String, Function], cellStyle: [Object, Function], headerRowClassName: [String, Function], headerRowStyle: [Object, Function], headerCellClassName: [String, Function], headerCellStyle: [Object, Function], highlightCurrentRow: Boolean, currentRowKey: [String, Number], emptyText: String, expandRowKeys: Array, defaultExpandAll: Boolean, defaultSort: Object, tooltipEffect: String, spanMethod: Function, selectOnIndeterminate: { type: Boolean, default: !0 }, indent: { type: Number, default: 16 }, treeProps: { type: Object, default: () => ({ hasChildren: "hasChildren", children: "children" }) }, lazy: Boolean, load: Function, style: { type: Object, default: () => ({}) }, className: { type: String, default: "" }, tableLayout: { type: String, default: "fixed" }, scrollbarAlwaysOn: { type: Boolean, default: !1 }, flexible: Boolean };
const Ae = { key: { type: String || Number, default: "0" }, columns: { type: Array, default: [] }, loading: { type: Boolean, default: !1 }, loadingConfig: { type: Object, default: () => {
} }, alignWhole: { type: String, default: "left" }, headerAlign: { type: String, default: "" }, showOverflowTooltip: { type: Boolean, default: !1 }, rowHoverBgColor: { type: String, default: "" }, pagination: { type: Object, default: { total: 0, pageSize: 5, align: "right", background: !1, pageSizes: [5, 10, 15, 20], layout: "total, sizes, prev, pager, next, jumper" } }, paginationSmall: { type: Boolean, default: !1 }, ...Pe }, fe = we({ name: "Renderer", props: { render: { type: Function }, params: { type: Object } }, setup: (e) => () => L(Oe, null, [e.render(e.params)]) }), he = Object.freeze(Object.defineProperty({ __proto__: null, default: {} }, Symbol.toStringTag, { value: "Module" }));
var Be = Object.create, ue = Object.defineProperty, ze = Object.getOwnPropertyDescriptor, Ee = Object.getOwnPropertyNames, We = Object.getPrototypeOf, Re = Object.prototype.hasOwnProperty, Se = (e, i) => () => (i || e((i = { exports: {} }).exports, i), i.exports), le = (e, i) => {
  for (var o in i)
    ue(e, o, { get: i[o], enumerable: !0 });
}, oe = (e, i, o, S) => {
  if (i && typeof i == "object" || typeof i == "function")
    for (let T of Ee(i))
      !Re.call(e, T) && T !== o && ue(e, T, { get: () => i[T], enumerable: !(S = ze(i, T)) || S.enumerable });
  return e;
}, ce = (e, i, o) => (oe(e, i, "default"), o && oe(o, i, "default")), Me = (e, i, o) => (o = e != null ? Be(We(e)) : {}, oe(!i && e && e.__esModule ? o : ue(o, "default", { value: e, enumerable: !0 }), e)), Le = Se((e, i) => {
  var o, S;
  o = e, S = function() {
    var T = 6e4, j = 36e5, Y = "millisecond", I = "second", _ = "minute", N = "hour", k = "day", H = "week", M = "month", Q = "quarter", D = "year", $ = "date", A = "Invalid Date", B = /^(\d{4})[-/]?(\d{1,2})?[-/]?(\d{0,2})[Tt\s]*(\d{1,2})?:?(\d{1,2})?:?(\d{1,2})?[.:]?(\d+)?$/, V = /\[([^\]]+)]|Y{1,4}|M{1,4}|D{1,2}|d{1,4}|H{1,2}|h{1,2}|a|A|m{1,2}|s{1,2}|Z{1,2}|SSS/g, F = { name: "en", weekdays: "Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday".split("_"), months: "January_February_March_April_May_June_July_August_September_October_November_December".split("_") }, E = function(a, r, t) {
      var c = String(a);
      return !c || c.length >= r ? a : "" + Array(r + 1 - c.length).join(t) + a;
    }, P = { s: E, z: function(a) {
      var r = -a.utcOffset(), t = Math.abs(r), c = Math.floor(t / 60), n = t % 60;
      return (r <= 0 ? "+" : "-") + E(c, 2, "0") + ":" + E(n, 2, "0");
    }, m: function a(r, t) {
      if (r.date() < t.date())
        return -a(t, r);
      var c = 12 * (t.year() - r.year()) + (t.month() - r.month()), n = r.clone().add(c, M), f = t - n < 0, p = r.clone().add(c + (f ? -1 : 1), M);
      return +(-(c + (t - n) / (f ? n - p : p - n)) || 0);
    }, a: function(a) {
      return a < 0 ? Math.ceil(a) || 0 : Math.floor(a);
    }, p: function(a) {
      return { M, y: D, w: H, d: k, D: $, h: N, m: _, s: I, ms: Y, Q }[a] || String(a || "").toLowerCase().replace(/s$/, "");
    }, u: function(a) {
      return a === void 0;
    } }, m = "en", h = {};
    h[m] = F;
    var u = function(a) {
      return a instanceof g;
    }, s = function a(r, t, c) {
      var n;
      if (!r)
        return m;
      if (typeof r == "string") {
        var f = r.toLowerCase();
        h[f] && (n = f), t && (h[f] = t, n = f);
        var p = r.split("-");
        if (!n && p.length > 1)
          return a(p[0]);
      } else {
        var y = r.name;
        h[y] = r, n = y;
      }
      return !c && n && (m = n), n || !c && m;
    }, d = function(a, r) {
      if (u(a))
        return a.clone();
      var t = typeof r == "object" ? r : {};
      return t.date = a, t.args = arguments, new g(t);
    }, l = P;
    l.l = s, l.i = u, l.w = function(a, r) {
      return d(a, { locale: r.$L, utc: r.$u, x: r.$x, $offset: r.$offset });
    };
    var g = function() {
      function a(t) {
        this.$L = s(t.locale, null, !0), this.parse(t);
      }
      var r = a.prototype;
      return r.parse = function(t) {
        this.$d = function(c) {
          var n = c.date, f = c.utc;
          if (n === null)
            return new Date(NaN);
          if (l.u(n))
            return new Date();
          if (n instanceof Date)
            return new Date(n);
          if (typeof n == "string" && !/Z$/i.test(n)) {
            var p = n.match(B);
            if (p) {
              var y = p[2] - 1 || 0, O = (p[7] || "0").substring(0, 3);
              return f ? new Date(Date.UTC(p[1], y, p[3] || 1, p[4] || 0, p[5] || 0, p[6] || 0, O)) : new Date(p[1], y, p[3] || 1, p[4] || 0, p[5] || 0, p[6] || 0, O);
            }
          }
          return new Date(n);
        }(t), this.$x = t.x || {}, this.init();
      }, r.init = function() {
        var t = this.$d;
        this.$y = t.getFullYear(), this.$M = t.getMonth(), this.$D = t.getDate(), this.$W = t.getDay(), this.$H = t.getHours(), this.$m = t.getMinutes(), this.$s = t.getSeconds(), this.$ms = t.getMilliseconds();
      }, r.$utils = function() {
        return l;
      }, r.isValid = function() {
        return this.$d.toString() !== A;
      }, r.isSame = function(t, c) {
        var n = d(t);
        return this.startOf(c) <= n && n <= this.endOf(c);
      }, r.isAfter = function(t, c) {
        return d(t) < this.startOf(c);
      }, r.isBefore = function(t, c) {
        return this.endOf(c) < d(t);
      }, r.$g = function(t, c, n) {
        return l.u(t) ? this[c] : this.set(n, t);
      }, r.unix = function() {
        return Math.floor(this.valueOf() / 1e3);
      }, r.valueOf = function() {
        return this.$d.getTime();
      }, r.startOf = function(t, c) {
        var n = this, f = !!l.u(c) || c, p = l.p(t), y = function(U, C) {
          var R = l.w(n.$u ? Date.UTC(n.$y, C, U) : new Date(n.$y, C, U), n);
          return f ? R : R.endOf(k);
        }, O = function(U, C) {
          return l.w(n.toDate()[U].apply(n.toDate("s"), (f ? [0, 0, 0, 0] : [23, 59, 59, 999]).slice(C)), n);
        }, w = this.$W, x = this.$M, W = this.$D, z = "set" + (this.$u ? "UTC" : "");
        switch (p) {
          case D:
            return f ? y(1, 0) : y(31, 11);
          case M:
            return f ? y(1, x) : y(0, x + 1);
          case H:
            var J = this.$locale().weekStart || 0, Z = (w < J ? w + 7 : w) - J;
            return y(f ? W - Z : W + (6 - Z), x);
          case k:
          case $:
            return O(z + "Hours", 0);
          case N:
            return O(z + "Minutes", 1);
          case _:
            return O(z + "Seconds", 2);
          case I:
            return O(z + "Milliseconds", 3);
          default:
            return this.clone();
        }
      }, r.endOf = function(t) {
        return this.startOf(t, !1);
      }, r.$set = function(t, c) {
        var n, f = l.p(t), p = "set" + (this.$u ? "UTC" : ""), y = (n = {}, n[k] = p + "Date", n[$] = p + "Date", n[M] = p + "Month", n[D] = p + "FullYear", n[N] = p + "Hours", n[_] = p + "Minutes", n[I] = p + "Seconds", n[Y] = p + "Milliseconds", n)[f], O = f === k ? this.$D + (c - this.$W) : c;
        if (f === M || f === D) {
          var w = this.clone().set($, 1);
          w.$d[y](O), w.init(), this.$d = w.set($, Math.min(this.$D, w.daysInMonth())).$d;
        } else
          y && this.$d[y](O);
        return this.init(), this;
      }, r.set = function(t, c) {
        return this.clone().$set(t, c);
      }, r.get = function(t) {
        return this[l.p(t)]();
      }, r.add = function(t, c) {
        var n, f = this;
        t = Number(t);
        var p = l.p(c), y = function(x) {
          var W = d(f);
          return l.w(W.date(W.date() + Math.round(x * t)), f);
        };
        if (p === M)
          return this.set(M, this.$M + t);
        if (p === D)
          return this.set(D, this.$y + t);
        if (p === k)
          return y(1);
        if (p === H)
          return y(7);
        var O = (n = {}, n[_] = T, n[N] = j, n[I] = 1e3, n)[p] || 1, w = this.$d.getTime() + t * O;
        return l.w(w, this);
      }, r.subtract = function(t, c) {
        return this.add(-1 * t, c);
      }, r.format = function(t) {
        var c = this, n = this.$locale();
        if (!this.isValid())
          return n.invalidDate || A;
        var f = t || "YYYY-MM-DDTHH:mm:ssZ", p = l.z(this), y = this.$H, O = this.$m, w = this.$M, x = n.weekdays, W = n.months, z = function(C, R, ne, G) {
          return C && (C[R] || C(c, f)) || ne[R].slice(0, G);
        }, J = function(C) {
          return l.s(y % 12 || 12, C, "0");
        }, Z = n.meridiem || function(C, R, ne) {
          var G = C < 12 ? "AM" : "PM";
          return ne ? G.toLowerCase() : G;
        }, U = { YY: String(this.$y).slice(-2), YYYY: this.$y, M: w + 1, MM: l.s(w + 1, 2, "0"), MMM: z(n.monthsShort, w, W, 3), MMMM: z(W, w), D: this.$D, DD: l.s(this.$D, 2, "0"), d: String(this.$W), dd: z(n.weekdaysMin, this.$W, x, 2), ddd: z(n.weekdaysShort, this.$W, x, 3), dddd: x[this.$W], H: String(y), HH: l.s(y, 2, "0"), h: J(1), hh: J(2), a: Z(y, O, !0), A: Z(y, O, !1), m: String(O), mm: l.s(O, 2, "0"), s: String(this.$s), ss: l.s(this.$s, 2, "0"), SSS: l.s(this.$ms, 3, "0"), Z: p };
        return f.replace(V, function(C, R) {
          return R || U[C] || p.replace(":", "");
        });
      }, r.utcOffset = function() {
        return 15 * -Math.round(this.$d.getTimezoneOffset() / 15);
      }, r.diff = function(t, c, n) {
        var f, p = l.p(c), y = d(t), O = (y.utcOffset() - this.utcOffset()) * T, w = this - y, x = l.m(this, y);
        return x = (f = {}, f[D] = x / 12, f[M] = x, f[Q] = x / 3, f[H] = (w - O) / 6048e5, f[k] = (w - O) / 864e5, f[N] = w / j, f[_] = w / T, f[I] = w / 1e3, f)[p] || w, n ? x : l.a(x);
      }, r.daysInMonth = function() {
        return this.endOf(M).$D;
      }, r.$locale = function() {
        return h[this.$L];
      }, r.locale = function(t, c) {
        if (!t)
          return this.$L;
        var n = this.clone(), f = s(t, c, !0);
        return f && (n.$L = f), n;
      }, r.clone = function() {
        return l.w(this.$d, this);
      }, r.toDate = function() {
        return new Date(this.valueOf());
      }, r.toJSON = function() {
        return this.isValid() ? this.toISOString() : null;
      }, r.toISOString = function() {
        return this.$d.toISOString();
      }, r.toString = function() {
        return this.$d.toUTCString();
      }, a;
    }(), b = g.prototype;
    return d.prototype = b, [["$ms", Y], ["$s", I], ["$m", _], ["$H", N], ["$W", k], ["$M", M], ["$y", D], ["$D", $]].forEach(function(a) {
      b[a[1]] = function(r) {
        return this.$g(r, a[0], a[1]);
      };
    }), d.extend = function(a, r) {
      return a.$i || (a(r, g, d), a.$i = !0), d;
    }, d.locale = s, d.isDayjs = u, d.unix = function(a) {
      return d(1e3 * a);
    }, d.en = h[m], d.Ls = h, d.p = {}, d;
  }, typeof e == "object" && typeof i < "u" ? i.exports = S() : typeof define == "function" && define.amd ? define(S) : (o = typeof globalThis < "u" ? globalThis : o || self).dayjs = S();
}), Qe = Se((e, i) => {
  var o, S;
  o = e, S = function() {
    var T, j, Y = 1e3, I = 6e4, _ = 36e5, N = 864e5, k = /\[([^\]]+)]|Y{1,4}|M{1,4}|D{1,2}|d{1,4}|H{1,2}|h{1,2}|a|A|m{1,2}|s{1,2}|Z{1,2}|SSS/g, H = 31536e6, M = 2592e6, Q = /^(-|\+)?P(?:([-+]?[0-9,.]*)Y)?(?:([-+]?[0-9,.]*)M)?(?:([-+]?[0-9,.]*)W)?(?:([-+]?[0-9,.]*)D)?(?:T(?:([-+]?[0-9,.]*)H)?(?:([-+]?[0-9,.]*)M)?(?:([-+]?[0-9,.]*)S)?)?$/, D = { years: H, months: M, days: N, hours: _, minutes: I, seconds: Y, milliseconds: 1, weeks: 6048e5 }, $ = function(h) {
      return h instanceof m;
    }, A = function(h, u, s) {
      return new m(h, s, u.$l);
    }, B = function(h) {
      return j.p(h) + "s";
    }, V = function(h) {
      return h < 0;
    }, F = function(h) {
      return V(h) ? Math.ceil(h) : Math.floor(h);
    }, E = function(h) {
      return Math.abs(h);
    }, P = function(h, u) {
      return h ? V(h) ? { negative: !0, format: "" + E(h) + u } : { negative: !1, format: "" + h + u } : { negative: !1, format: "" };
    }, m = function() {
      function h(s, d, l) {
        var g = this;
        if (this.$d = {}, this.$l = l, s === void 0 && (this.$ms = 0, this.parseFromMilliseconds()), d)
          return A(s * D[B(d)], this);
        if (typeof s == "number")
          return this.$ms = s, this.parseFromMilliseconds(), this;
        if (typeof s == "object")
          return Object.keys(s).forEach(function(r) {
            g.$d[B(r)] = s[r];
          }), this.calMilliseconds(), this;
        if (typeof s == "string") {
          var b = s.match(Q);
          if (b) {
            var a = b.slice(2).map(function(r) {
              return r != null ? Number(r) : 0;
            });
            return this.$d.years = a[0], this.$d.months = a[1], this.$d.weeks = a[2], this.$d.days = a[3], this.$d.hours = a[4], this.$d.minutes = a[5], this.$d.seconds = a[6], this.calMilliseconds(), this;
          }
        }
        return this;
      }
      var u = h.prototype;
      return u.calMilliseconds = function() {
        var s = this;
        this.$ms = Object.keys(this.$d).reduce(function(d, l) {
          return d + (s.$d[l] || 0) * D[l];
        }, 0);
      }, u.parseFromMilliseconds = function() {
        var s = this.$ms;
        this.$d.years = F(s / H), s %= H, this.$d.months = F(s / M), s %= M, this.$d.days = F(s / N), s %= N, this.$d.hours = F(s / _), s %= _, this.$d.minutes = F(s / I), s %= I, this.$d.seconds = F(s / Y), s %= Y, this.$d.milliseconds = s;
      }, u.toISOString = function() {
        var s = P(this.$d.years, "Y"), d = P(this.$d.months, "M"), l = +this.$d.days || 0;
        this.$d.weeks && (l += 7 * this.$d.weeks);
        var g = P(l, "D"), b = P(this.$d.hours, "H"), a = P(this.$d.minutes, "M"), r = this.$d.seconds || 0;
        this.$d.milliseconds && (r += this.$d.milliseconds / 1e3);
        var t = P(r, "S"), c = s.negative || d.negative || g.negative || b.negative || a.negative || t.negative, n = b.format || a.format || t.format ? "T" : "", f = (c ? "-" : "") + "P" + s.format + d.format + g.format + n + b.format + a.format + t.format;
        return f === "P" || f === "-P" ? "P0D" : f;
      }, u.toJSON = function() {
        return this.toISOString();
      }, u.format = function(s) {
        var d = s || "YYYY-MM-DDTHH:mm:ss", l = { Y: this.$d.years, YY: j.s(this.$d.years, 2, "0"), YYYY: j.s(this.$d.years, 4, "0"), M: this.$d.months, MM: j.s(this.$d.months, 2, "0"), D: this.$d.days, DD: j.s(this.$d.days, 2, "0"), H: this.$d.hours, HH: j.s(this.$d.hours, 2, "0"), m: this.$d.minutes, mm: j.s(this.$d.minutes, 2, "0"), s: this.$d.seconds, ss: j.s(this.$d.seconds, 2, "0"), SSS: j.s(this.$d.milliseconds, 3, "0") };
        return d.replace(k, function(g, b) {
          return b || String(l[g]);
        });
      }, u.as = function(s) {
        return this.$ms / D[B(s)];
      }, u.get = function(s) {
        var d = this.$ms, l = B(s);
        return l === "milliseconds" ? d %= 1e3 : d = l === "weeks" ? F(d / D[l]) : this.$d[l], d === 0 ? 0 : d;
      }, u.add = function(s, d, l) {
        var g;
        return g = d ? s * D[B(d)] : $(s) ? s.$ms : A(s, this).$ms, A(this.$ms + g * (l ? -1 : 1), this);
      }, u.subtract = function(s, d) {
        return this.add(s, d, !0);
      }, u.locale = function(s) {
        var d = this.clone();
        return d.$l = s, d;
      }, u.clone = function() {
        return A(this.$ms, this);
      }, u.humanize = function(s) {
        return T().add(this.$ms, "ms").locale(this.$l).fromNow(!s);
      }, u.milliseconds = function() {
        return this.get("milliseconds");
      }, u.asMilliseconds = function() {
        return this.as("milliseconds");
      }, u.seconds = function() {
        return this.get("seconds");
      }, u.asSeconds = function() {
        return this.as("seconds");
      }, u.minutes = function() {
        return this.get("minutes");
      }, u.asMinutes = function() {
        return this.as("minutes");
      }, u.hours = function() {
        return this.get("hours");
      }, u.asHours = function() {
        return this.as("hours");
      }, u.days = function() {
        return this.get("days");
      }, u.asDays = function() {
        return this.as("days");
      }, u.weeks = function() {
        return this.get("weeks");
      }, u.asWeeks = function() {
        return this.as("weeks");
      }, u.months = function() {
        return this.get("months");
      }, u.asMonths = function() {
        return this.as("months");
      }, u.years = function() {
        return this.get("years");
      }, u.asYears = function() {
        return this.as("years");
      }, h;
    }();
    return function(h, u, s) {
      T = s, j = s().$utils(), s.duration = function(g, b) {
        var a = s.locale();
        return A(g, { $l: a }, b);
      }, s.isDuration = $;
      var d = u.prototype.add, l = u.prototype.subtract;
      u.prototype.add = function(g, b) {
        return $(g) && (g = g.asMilliseconds()), d.bind(this)(g, b);
      }, u.prototype.subtract = function(g, b) {
        return $(g) && (g = g.asMilliseconds()), l.bind(this)(g, b);
      };
    };
  }, typeof e == "object" && typeof i < "u" ? i.exports = S() : typeof define == "function" && define.amd ? define(S) : (o = typeof globalThis < "u" ? globalThis : o || self).dayjs_plugin_duration = S();
});
function pe(e) {
  return typeof e == "function";
}
var Ve = Me(Le()), Ue = Me(Qe());
Ve.default.extend(Ue.default);
var me = {};
le(me, { Fs: () => he }), ce(me, he);
var ge = class {
  constructor(e) {
    de(this, "storage");
    this.storage = e;
  }
  setItem(e, i) {
    this.storage.setItem(e, JSON.stringify(i));
  }
  getItem(e) {
    return JSON.parse(this.storage.getItem(e));
  }
  removeItem(e) {
    this.storage.removeItem(e);
  }
  clear() {
    this.storage.clear();
  }
}, Je = typeof window < "u" ? window.localStorage : "", Ze = typeof window < "u" ? window.sessionStorage : "";
new class extends ge {
  constructor(e) {
    super(e);
  }
}(Je), new ge(Ze);
var ee = {};
le(ee, { Vue: () => X }), ce(ee, X);
var te = {};
le(te, { Vue: () => X, Vue2: () => qe, del: () => tt, install: () => Xe, isVue2: () => Ke, isVue3: () => Ge, set: () => et }), ce(te, X);
var ye, Ke = !1, Ge = !0, qe = void 0;
function Xe() {
}
function et(e, i, o) {
  return Array.isArray(e) ? (e.length = Math.max(e.length, i), e.splice(i, 1, o), o) : (e[i] = o, o);
}
function tt(e, i) {
  Array.isArray(e) ? e.splice(i, 1) : delete e[i];
}
typeof window < "u" && ((ye = window == null ? void 0 : window.navigator) == null ? void 0 : ye.userAgent) && /iP(ad|hone|od)/.test(window.navigator.userAgent);
var $e, K, se = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, ie = "__vueuse_ssr_handlers__";
se[ie] = se[ie] || {}, se[ie], (K = $e || ($e = {})).UP = "UP", K.RIGHT = "RIGHT", K.DOWN = "DOWN", K.LEFT = "LEFT", K.NONE = "NONE";
var nt = Object.defineProperty, ve = Object.getOwnPropertySymbols, rt = Object.prototype.hasOwnProperty, st = Object.prototype.propertyIsEnumerable, be = (e, i, o) => i in e ? nt(e, i, { enumerable: !0, configurable: !0, writable: !0, value: o }) : e[i] = o;
((e, i) => {
  for (var o in i || (i = {}))
    rt.call(i, o) && be(e, o, i[o]);
  if (ve)
    for (var o of ve(i))
      st.call(i, o) && be(e, o, i[o]);
})({ linear: function(e) {
  return e;
} }, { easeInSine: [0.12, 0, 0.39, 0], easeOutSine: [0.61, 1, 0.88, 1], easeInOutSine: [0.37, 0, 0.63, 1], easeInQuad: [0.11, 0, 0.5, 0], easeOutQuad: [0.5, 1, 0.89, 1], easeInOutQuad: [0.45, 0, 0.55, 1], easeInCubic: [0.32, 0, 0.67, 0], easeOutCubic: [0.33, 1, 0.68, 1], easeInOutCubic: [0.65, 0, 0.35, 1], easeInQuart: [0.5, 0, 0.75, 0], easeOutQuart: [0.25, 1, 0.5, 1], easeInOutQuart: [0.76, 0, 0.24, 1], easeInQuint: [0.64, 0, 0.78, 0], easeOutQuint: [0.22, 1, 0.36, 1], easeInOutQuint: [0.83, 0, 0.17, 1], easeInExpo: [0.7, 0, 0.84, 0], easeOutExpo: [0.16, 1, 0.3, 1], easeInOutExpo: [0.87, 0, 0.13, 1], easeInCirc: [0.55, 0, 1, 0.45], easeOutCirc: [0, 0.55, 0.45, 1], easeInOutCirc: [0.85, 0, 0.15, 1], easeInBack: [0.36, 0, 0.66, -0.56], easeOutBack: [0.34, 1.56, 0.64, 1], easeInOutBack: [0.68, -0.6, 0.32, 1.6] });
var it = () => {
  let e, i = (0, ee.ref)(!1), o = () => {
    i.value = document.documentElement.classList.contains("dark");
  };
  return function(S) {
    (0, te.getCurrentInstance)() && (0, te.onUnmounted)(S);
  }(() => {
    e.takeRecords(), e.disconnect();
  }), (0, ee.onBeforeMount)(() => {
    o(), e = new MutationObserver(o), e.observe(document.documentElement, { attributes: !0, attributeFilter: ["class"] });
  }), { isDark: i };
};
const ae = we({ name: "PureTable", props: Ae, emits: ["size-change", "current-change"], setup(e, { slots: i, attrs: o, emit: S, expose: T }) {
  const { isDark: j } = it(), Y = je();
  function I() {
    var m;
    return (m = Y == null ? void 0 : Y.proxy) == null ? void 0 : m.$refs[`TableRef${e.key}`];
  }
  function _() {
    return I().$refs;
  }
  Te(() => {
    Ce(() => {
      e.rowHoverBgColor && _().tableWrapper.style.setProperty("--el-table-row-hover-bg-color", e.rowHoverBgColor, "important");
    });
  });
  const { columns: N, loading: k, loadingConfig: H, alignWhole: M, headerAlign: Q, showOverflowTooltip: D, pagination: $ } = Ie(e);
  let A = re(() => {
    if (!v(H))
      return;
    let { text: m, spinner: h, svg: u, viewBox: s } = v(H);
    return { "element-loading-text": m, "element-loading-spinner": h, "element-loading-svg": u, "element-loading-svg-view-box": s };
  });
  const B = re(() => {
    var m, h;
    return ((h = (m = v(H)) == null ? void 0 : m.background) != null ? h : j.value) ? "rgba(0, 0, 0, 0.45)" : "rgba(255, 255, 255, 0.45)";
  }), V = re(() => {
    var m;
    return Object.assign({ width: "100%", margin: "16px 0", display: "flex", justifyContent: v($).align === "left" ? "flex-start" : v($).align === "center" ? "center" : "flex-end" }, (m = v($).style) != null ? m : {});
  });
  let F = v($) && v($).currentPage && v($).pageSize;
  const E = (m, h) => {
    const { cellRenderer: u, slot: s, headerRenderer: d, hide: l, children: g, prop: b, ...a } = m, r = { default: (n) => {
      var f;
      return u ? L(fe, { render: u, params: Object.assign(n, { index: n.$index, props: e, attrs: o }) }, null) : s ? (f = i == null ? void 0 : i[s]) == null ? void 0 : f.call(i, Object.assign(n, { index: n.$index, props: e, attrs: o })) : void 0;
    } };
    let t = d ? { header: (n) => L(fe, { render: d, params: Object.assign(n, { index: n.$index, props: e, attrs: o }) }, null), ...r } : r;
    return pe(l) && l(o) ? l(o) : ((g == null ? void 0 : g.length) > 0 && (t = g.map(E)), L(Fe, q({ key: h }, a, { prop: pe(b) && b(h) ? b(h) : b, align: m != null && m.align ? m.align : v(M), headerAlign: m != null && m.headerAlign ? m.headerAlign : v(Q), showOverflowTooltip: m != null && m.showOverflowTooltip ? m.showOverflowTooltip : v(D) }), typeof (c = t) == "function" || Object.prototype.toString.call(c) === "[object Object]" && !Ne(c) ? t : { default: () => [t] }));
    var c;
  };
  T({ getTableRef: I, getTableDoms: _ });
  let P = () => {
    var m, h;
    return L(Oe, null, [L(He, q(e, o, { ref: `TableRef${e.key}` }), { default: () => v(N).map(E), append: () => i.append && i.append(), empty: () => i.empty && i.empty() }), F ? L(Ye, q(o, { class: "pure-pagination", style: v(V) }, v($), { small: e != null && e.paginationSmall ? e == null ? void 0 : e.paginationSmall : !!v($).small && v($).small, layout: (m = v($).layout) != null ? m : "total, sizes, prev, pager, next, jumper", pageSizes: (h = v($).pageSizes) != null ? h : [5, 10, 15, 20], onSizeChange: (u) => ((s) => {
      v($).pageSize = s, S("size-change", s);
    })(u), onCurrentChange: (u) => ((s) => {
      v($).currentPage = s, S("current-change", s);
    })(u) }), null) : null]);
  };
  return () => v(k) ? _e(L("div", q({ class: "pure-table", "element-loading-background": v(B) }, v(A)), [P()]), [[ke("loading"), v(k)]]) : P();
} }), lt = Object.assign(ae, { install: function(e) {
  e.component(ae.name, ae);
} });
export {
  lt as PureTable,
  lt as default
};
