// vxe-table@4.3.6/es/loading/src/loading.js
import { defineComponent, h, computed } from 'vue';
import GlobalConfig from '../../v-x-e-table/src/conf';
export default defineComponent({
    name: 'VxeLoading',
    props: {
        modelValue: Boolean,
        icon: String,
        text: String
    },
    setup: function (props) {
        var computeLoadingIcon = computed(function () {
            return props.icon || GlobalConfig.icon.LOADING;
        });
        var computeLoadingText = computed(function () {
            var loadingText = GlobalConfig.loadingText;
            return props.text || (loadingText === null ? loadingText : GlobalConfig.i18n('vxe.loading.text'));
        });
        return function () {
            var loadingIcon = computeLoadingIcon.value;
            var loadingText = computeLoadingText.value;
            return h('div', {
                class: ['vxe-loading', {
                        'is--visible': props.modelValue
                    }]
            }, [
                h('div', {
                    class: 'vxe-loading--chunk'
                }, [
                    loadingIcon ? h('i', {
                        class: loadingIcon
                    }) : h('div', {
                        class: 'vxe-loading--spinner'
                    }),
                    loadingText ? h('div', {
                        class: 'vxe-loading--text'
                    }, "".concat(loadingText)) : null
                ])
            ]);
        };
    }
});
