// vxe-table@4.3.6/lib/v-x-e-table/src/store.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Store = void 0;
var _log = require("../../tools/log");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/**
 * 创建数据仓库
 */
var Store = /** @class */function () {
  function Store() {
    this.store = {};
  }
  Store.prototype.mixin = function (options) {
    Object.assign(this.store, options);
    return this;
  };
  Store.prototype.has = function (name) {
    return !!this.get(name);
  };
  Store.prototype.get = function (name) {
    return this.store[name];
  };
  Store.prototype.add = function (name, render) {
    var conf = this.store[name];
    // 检测是否覆盖
    if (process.env.NODE_ENV === 'development') {
      var confKeys_1 = _xeUtils.default.keys(conf);
      _xeUtils.default.each(render, function (item, key) {
        if (confKeys_1.includes(key)) {
          (0, _log.warnLog)('vxe.error.coverProp', [name, key]);
        }
      });
    }
    this.store[name] = conf ? _xeUtils.default.merge(conf, render) : render;
    return this;
  };
  Store.prototype.delete = function (name) {
    delete this.store[name];
  };
  Store.prototype.forEach = function (callback) {
    _xeUtils.default.objectEach(this.store, callback);
  };
  return Store;
}();
exports.Store = Store;
var _default = Store;
exports.default = _default;
