// @vue-macros/common@0.13.4/dist/chunk-HNRHZMFS.mjs
// src/constants.ts
var DEFINE_PROPS = "defineProps";
var DEFINE_PROPS_DOLLAR = "$defineProps";
var DEFINE_EMITS = "defineEmits";
var WITH_DEFAULTS = "withDefaults";
var DEFINE_OPTIONS = "defineOptions";
var DEFINE_MODEL = "defineModel";
var DEFINE_MODEL_DOLLAR = "$defineModel";
var DEFINE_SETUP_COMPONENT = "defineSetupComponent";
var DEFINE_RENDER = "defineRender";
var DEFINE_SLOTS = "defineSlots";
var REPO_ISSUE_URL = "https://github.com/sxzz/unplugin-vue-macros/issues";
var REGEX_SRC_FILE = /\.[cm]?[jt]sx?$/;
var REGEX_TS_FILE = /^[cm]?tsx?$/;
var REGEX_JSX_FILE = /^[cm]?[jt]sx$/;
var REGEX_SETUP_SFC = /\.setup\.[cm]?[jt]sx?$/;
var REGEX_VUE_SFC = /\.vue$/;
var REGEX_VUE_SUB = /\.vue\?vue&/;

export {
  DEFINE_PROPS,
  DEFINE_PROPS_DOLLAR,
  DEFINE_EMITS,
  WITH_DEFAULTS,
  DEFINE_OPTIONS,
  DEFINE_MODEL,
  DEFINE_MODEL_DOLLAR,
  DEFINE_SETUP_COMPONENT,
  DEFINE_RENDER,
  DEFINE_SLOTS,
  REPO_ISSUE_URL,
  REGEX_SRC_FILE,
  REGEX_TS_FILE,
  REGEX_JSX_FILE,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB
};
