// @vue-macros/common@0.13.4/dist/chunk-UDPJYA4D.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/constants.ts
var DEFINE_PROPS = "defineProps";
var DEFINE_PROPS_DOLLAR = "$defineProps";
var DEFINE_EMITS = "defineEmits";
var WITH_DEFAULTS = "withDefaults";
var DEFINE_OPTIONS = "defineOptions";
var DEFINE_MODEL = "defineModel";
var DEFINE_MODEL_DOLLAR = "$defineModel";
var DEFINE_SETUP_COMPONENT = "defineSetupComponent";
var DEFINE_RENDER = "defineRender";
var DEFINE_SLOTS = "defineSlots";
var REPO_ISSUE_URL = "https://github.com/sxzz/unplugin-vue-macros/issues";
var REGEX_SRC_FILE = /\.[cm]?[jt]sx?$/;
var REGEX_TS_FILE = /^[cm]?tsx?$/;
var REGEX_JSX_FILE = /^[cm]?[jt]sx$/;
var REGEX_SETUP_SFC = /\.setup\.[cm]?[jt]sx?$/;
var REGEX_VUE_SFC = /\.vue$/;
var REGEX_VUE_SUB = /\.vue\?vue&/;



















exports.DEFINE_PROPS = DEFINE_PROPS; exports.DEFINE_PROPS_DOLLAR = DEFINE_PROPS_DOLLAR; exports.DEFINE_EMITS = DEFINE_EMITS; exports.WITH_DEFAULTS = WITH_DEFAULTS; exports.DEFINE_OPTIONS = DEFINE_OPTIONS; exports.DEFINE_MODEL = DEFINE_MODEL; exports.DEFINE_MODEL_DOLLAR = DEFINE_MODEL_DOLLAR; exports.DEFINE_SETUP_COMPONENT = DEFINE_SETUP_COMPONENT; exports.DEFINE_RENDER = DEFINE_RENDER; exports.DEFINE_SLOTS = DEFINE_SLOTS; exports.REPO_ISSUE_URL = REPO_ISSUE_URL; exports.REGEX_SRC_FILE = REGEX_SRC_FILE; exports.REGEX_TS_FILE = REGEX_TS_FILE; exports.REGEX_JSX_FILE = REGEX_JSX_FILE; exports.REGEX_SETUP_SFC = REGEX_SETUP_SFC; exports.REGEX_VUE_SFC = REGEX_VUE_SFC; exports.REGEX_VUE_SUB = REGEX_VUE_SUB;
