// vxe-table@4.3.6/lib/loading/src/loading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _conf = _interopRequireDefault(require("../../v-x-e-table/src/conf"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = (0, _vue.defineComponent)({
  name: 'VxeLoading',
  props: {
    modelValue: Boolean,
    icon: String,
    text: String
  },
  setup: function (props) {
    var computeLoadingIcon = (0, _vue.computed)(function () {
      return props.icon || _conf.default.icon.LOADING;
    });
    var computeLoadingText = (0, _vue.computed)(function () {
      var loadingText = _conf.default.loadingText;
      return props.text || (loadingText === null ? loadingText : _conf.default.i18n('vxe.loading.text'));
    });
    return function () {
      var loadingIcon = computeLoadingIcon.value;
      var loadingText = computeLoadingText.value;
      return (0, _vue.h)('div', {
        class: ['vxe-loading', {
          'is--visible': props.modelValue
        }]
      }, [(0, _vue.h)('div', {
        class: 'vxe-loading--chunk'
      }, [loadingIcon ? (0, _vue.h)('i', {
        class: loadingIcon
      }) : (0, _vue.h)('div', {
        class: 'vxe-loading--spinner'
      }), loadingText ? (0, _vue.h)('div', {
        class: 'vxe-loading--text'
      }, "".concat(loadingText)) : null])]);
    };
  }
});
exports.default = _default;
