// @vue-macros/api@0.2.0/dist/ts.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
















var _chunk6EAABA32js = require('./chunk-6EAABA32.js');

















exports.exportsSymbol = _chunk6EAABA32js.exportsSymbol; exports.getTSFile = _chunk6EAABA32js.getTSFile; exports.isTSDeclaration = _chunk6EAABA32js.isTSDeclaration; exports.isTSExports = _chunk6EAABA32js.isTSExports; exports.mergeTSProperties = _chunk6EAABA32js.mergeTSProperties; exports.resolveTSEntityName = _chunk6EAABA32js.resolveTSEntityName; exports.resolveTSExports = _chunk6EAABA32js.resolveTSExports; exports.resolveTSFileId = _chunk6EAABA32js.resolveTSFileId; exports.resolveTSFileIdNode = _chunk6EAABA32js.resolveTSFileIdNode; exports.resolveTSProperties = _chunk6EAABA32js.resolveTSProperties; exports.resolveTSReferencedType = _chunk6EAABA32js.resolveTSReferencedType; exports.resolveTSScope = _chunk6EAABA32js.resolveTSScope; exports.resolveTypeElements = _chunk6EAABA32js.resolveTypeElements; exports.setResolveTSFileIdImpl = _chunk6EAABA32js.setResolveTSFileIdImpl; exports.tsFileCache = _chunk6EAABA32js.tsFileCache; exports.tsFileExportsCache = _chunk6EAABA32js.tsFileExportsCache;
exports.default = module.exports;
